docs/features/software-templates/writing-custom-filters.md

```diff
@@ -0,0 +1,145 @@
+  useEntityList,
+} from '@backstage/plugin-catalog-react';
+import { Entity } from '@backstage/catalog-model';
+import Checkbox from '@material-ui/core/Checkbox';
+import FormControl from '@material-ui/core/FormControl';
+import FormControlLabel from '@material-ui/core/FormControlLabel';
+import FormGroup from '@material-ui/core/FormGroup';
+import Typography from '@material-ui/core/Typography';
+
+export type EntityTagPickerProps = {
+  label: string;
+  allowedTags: string[];
+  showCounts?: boolean;
+};
+
+class CustomEntityTagFilter implements EntityFilter {
+  constructor(readonly values: string[]) {}
+
+  filterEntity(entity: Entity): boolean {
+    const entityTags = entity.metadata.tags ?? [];
+    return this.values.every(tag => entityTags.includes(tag));
+  }
+}
+
+export type CustomFilters = DefaultEntityFilters & {
+  customTags: CustomEntityTagFilter;
+};
+
+export const CustomEntityTagPicker = (props: EntityTagPickerProps) => {
+  const {
+    filters: { customTags },
+    updateFilters,
+  } = useEntityList<CustomFilters>();
+
+  function onChange(value: string) {
+    const newTags = customTags?.values.includes(value)
+      ? customTags.values.filter(tag => tag !== value)
+      : [...(customTags?.values ?? []), value];
+    updateFilters({
+      customTags: newTags.length
+        ? new CustomEntityTagFilter(newTags)
+        : undefined,
+    });
+  }
+  const allowedTags = props.allowedTags.map(tag =>
+    tag.toLocaleLowerCase('en-US'),
+  );
+
+  return (
+    <Box display="flex" flexDirection="column">
+      <FormControl component="fieldset">
+        <Typography variant="button">{props.label}</Typography>
+        <FormGroup>
+          {allowedTags.map(tier => (
+            <FormControlLabel
+              key={tier}
+              control={
+                <Checkbox
+                  checked={customTags?.values.includes(tier)}
+                  onChange={() => onChange(tier)}
+                />
+              }
+              label={`${tier.charAt(0).toLocaleUpperCase('en-US')}${tier.slice(
+                1,
+              )}`}
+            />
+          ))}
+        </FormGroup>
+      </FormControl>
+    </Box>
+  );
+};
+```
+
+Once we have the `CustomEntityTagPicker` component, we can use it in `App.tsx` to override the default filters. We can also add other filters like `EntityTagPicker` and `EntityOwnerPicker` to the list of filters.
+
+:::note Note
+
+The `EntityKindPicker` is required in order to display all the existing templates. It is hidden by default in order to improve the user experience.
+
+:::
+
+```tsx title="src/App.tsx"
+// ...
+<Route
+  path="/create"
+  element={
+    <ScaffolderPage
+      defaultPreviewTemplate={defaultPreviewTemplate}
+      groups={[
+        {
+          title: 'Recommended',
+          filter: entity =>
+            entity?.metadata?.tags?.includes('recommended') ?? false,
+        },
+      ]}
+      {/* highlight-add-start */}
+      overrideFilters={
```

Comment:
I just modified this so that we can pass in a custom component through the body of the plugin. However, this requires the component to be registered using the `createScaffolderFilter` function (the same way the layout/extensions work). Furthermore, the `TemplateListPage` still uses an `additionalFilters` prop to get the filters from the scaffolder Router via the `useCustomFilters` hook.

Let me know if this makes sense or not! 😃